pkg/arvo/app/launch/js/channel.js

```diff
@@ -196,8 +210,9 @@ class Channel {
     }
 
     this.eventSource.onerror = e => {
-      console.error("eventSource error:", e);
       this.delete();
+      this.init();
+      this.onChannelError(e);
```

Comment:
Uh, I thought the channel design was made so that you can reestablish a connection with the _same_ channel after error, which avoids opening new subscriptions on the backend, and thus doesn't get you superfluous initial results. Doing a full delete and init here seems to run counter to that.

(And even if this is sane, you might want to call `onChannelError` before doing all of that, for the expected order of operations.)

I'd expect the best behavior here to be something along the lines of checking `eventSource.readyState`, and reestablishing the connection (maybe through `connectIfDisconnected`) if that's `EventSource.CLOSED`. You probably don't want to throw away client-side state. This could lose you poke n/ack handling, forces you to reestablish subscriptions, etc.

I'm not yet as familiar with this as I'd like to be, so I'm only 80% sure. @eglaysher care to take a look to confirm/deny my suspicions?

Reply:
No, if the channel failed, I think you want to treat it just like you would a page refresh. You need the "superfluous" initial results to receive any information you lost while the eventSource was down. You want to run `onChannelError` after a new channel has been initialized so that you can be sure that you have a new Channel that you can work with to re-establish subscriptions, since this one will have been cleaned up already.

Reply:
Makes sense re call order. Maybe `onChannelReconnect` or similar would be more clear then.

> You need the "superfluous" initial results to receive any information you lost while the eventSource was down.

I was under the impression channel.js actively acks events as they come in, enabling Eyre to hold on to events we haven't received, to then send those on reconnect.

https://github.com/urbit/urbit/blob/cc16e905afda468afe0cebd36e1a713fb9f6fd74/pkg/arvo/sys/vane/eyre.hoon#L233-L241

Reply:
Nope. If you look at the TODO in +channel-js in Eyre, it says: "Must send acks to the server"

Reply:
This issue is separate from the current PR. The current PR is ready to be merged.

Reply:
> Nope. If you look at the TODO in +channel-js in Eyre, it says: "Must send acks to the server"

Oh, you're right. Agreed it's not _strictly_ part of "reconnect on errors", but not/having that todo in heavily influences implementation and behavior here.

Seems to me like we should make an action item for implementing that todo, and then (assuming this gets merged) revisit the work here to account for that actually working as intended. I'll make sure this gets noted somewhere.

Reply:
Thanks, I completely agree.